chain-impl-mockchain/src/tokens.rs

```diff
@@ -0,0 +1,208 @@
+    pub policy_hash: PolicyHash,
+    pub token_name: TokenName,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PolicyHash([u8; POLICY_HASH_SIZE]);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TokenName(Vec<u8>);
+
+/// A minting policy consists of multiple entries defining different
+/// constraints on the minting process. An empty policy means that new tokens
+/// cannot be minted during the chain run.
+///
+/// Minting policies are meant to be ignored in block0 fragments.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MintingPolicy(Vec<MintingPolicyEntry>);
+
+/// An entry of a minting policy. Currently there are no entries available.
+/// This is reserved for the future use.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MintingPolicyEntry {}
+
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+pub enum MintingPolicyViolation {
+    #[error("the policy of this token does not allow minting")]
+    AdditionalMintingNotAllowed,
+}
+
+#[derive(Debug, Error)]
+#[error("Token name can be no more that {} bytes long; got {} bytes", TOKEN_NAME_MAX_BYTES, .actual)]
+pub struct TokenNameTooLong {
+    actual: usize,
+}
+
+impl MintingPolicy {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn check_minting_tx(&self) -> Result<(), MintingPolicyViolation> {
+        if self.0.is_empty() {
+            return Err(MintingPolicyViolation::AdditionalMintingNotAllowed);
+        }
+
+        for _entry in &self.0 {
+            unreachable!("implement this when we have actual minting policies");
+        }
+
+        Ok(())
+    }
+
+    pub fn entries(&self) -> &[MintingPolicyEntry] {
+        &self.0
+    }
+
+    pub fn bytes(&self) -> Vec<u8> {
+        let bb: ByteBuilder<Self> = ByteBuilder::new();
+        bb.u8(0).finalize_as_vec()
+    }
+
+    pub fn hash(&self) -> PolicyHash {
+        let mut result = [0u8; POLICY_HASH_SIZE];
+        if self.0.is_empty() {
+            let mut hasher = Blake2b::new(POLICY_HASH_SIZE);
+            hasher.input(&self.bytes());
+            hasher.result(&mut result);
+        }
+        PolicyHash(result)
+    }
+}
+
+impl Default for MintingPolicy {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Readable for MintingPolicy {
+    fn read(buf: &mut ReadBuf) -> Result<Self, ReadError> {
+        let no_entries = buf.get_u8()?;
+        if no_entries != 0 {
+            return Err(ReadError::InvalidData(
+                "non-zero number of minting policy entries, but they are currently unimplemented"
+                    .to_string(),
+            ));
+        }
+        Ok(Self::new())
+    }
+}
+
+impl PolicyHash {
+    pub fn bytes(&self) -> &[u8; POLICY_HASH_SIZE] {
+        &self.0
+    }
+}
+
+impl Readable for PolicyHash {
+    fn read(buf: &mut ReadBuf) -> Result<Self, ReadError> {
+        let bytes = buf
+            .get_slice(POLICY_HASH_SIZE)?
+            .try_into()
+            .expect("already read 32 bytes");
```

Comment:
what's up with this error?

Reply:
Here we convert `&[u8]` to `[u8; 32]`. This conversion can fail but the only condition for this would be size mismatch. It should not be possible if our code is functioning correctly, so I put an `.expect()` here to catch some severe misbehaviour.

Reply:
But isn't the policy hash 28 bytes?

Reply:
Oopsie, should've used `format!()` here maybe.